Sources/SwiftOperatorPrecedence/OperatorPrecedence+Folding.swift

```diff
@@ -0,0 +1,364 @@
+  /// as if the expression has been parenthesized `x + (y * z)`.
+  public func foldSingle(
+    _ sequence: SequenceExprSyntax,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> ExprSyntax {
+    let lhs = sequence.elements.first!
+    var rest = sequence.elements.dropFirst()
+    return try fold(
+      lhs, rest: &rest,
+      bound: PrecedenceBound(groupName: nil, isStrict: false, syntax: nil),
+      errorHandler: errorHandler
+    )
+  }
+
+  /// Syntax rewriter that folds all of the sequence expressions it
+  /// encounters.
+  private class SequenceFolder : SyntaxRewriter {
+    /// The first operator precedecence that caused the error handler to
+    /// also throw.
+    var firstFatalError: OperatorPrecedenceError? = nil
+
+    let opPrecedence: OperatorPrecedence
+    let errorHandler: OperatorPrecedenceErrorHandler
+
+    init(
+      opPrecedence: OperatorPrecedence,
+      errorHandler: @escaping OperatorPrecedenceErrorHandler
+    ) {
+      self.opPrecedence = opPrecedence
+      self.errorHandler = errorHandler
+    }
+
+    override func visit(_ node: SequenceExprSyntax) -> ExprSyntax {
+      // If the error handler threw in response to an error, don't
+      // continue folding.
+      if firstFatalError != nil {
+        return ExprSyntax(node)
+      }
+
+      let newNode = super.visit(node).as(SequenceExprSyntax.self)!
+
+      // If the error handler threw in response to an error, don't
+      // continue folding.
+      if firstFatalError != nil {
+        return ExprSyntax(newNode)
+      }
+
+      // Try to fold this sequence expression.
+      do {
+        return try opPrecedence.foldSingle(newNode) { origError in
+          do {
+            try errorHandler(origError)
+          } catch {
+            firstFatalError = origError
+            throw error
+          }
+        }
+      } catch {
+        return ExprSyntax(newNode)
+      }
+    }
+  }
+
+  /// Fold all sequence expressions within the given syntax tree into a
+  /// structured syntax tree.
+  ///
+  /// This operation replaces all sequence expressions in the given syntax
+  /// tree with structured syntax trees, by walking the tree and invoking
+  /// `foldSingle` on each sequence expression it encounters. Use this to
+  /// provide structure to an entire tree.
+  public func foldAll<Node: SyntaxProtocol>(
+    _ node: Node,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> SyntaxProtocol {
+    return try withoutActuallyEscaping(errorHandler) { errorHandler in
+      let folder = SequenceFolder(
+        opPrecedence: self, errorHandler: errorHandler
+      )
+      let result = folder.visit(Syntax(node))
+
+      // If the sequence folder encountered an error that caused the error
+      // handler to throw, invoke the error handler again with the original
+      // error.
+      if let origFatalError = folder.firstFatalError {
+        try errorHandler(origFatalError)
```

Comment:
Doesn’t this call `errorHandler` twice? Once from `SequenceFolder` and once from here?

Reply:
Yeah, the problem here is that `rethrows` checking cannot account for errors thrown in the visitor. I'll document this as a limitation. The real fix is something like https://github.com/apple/swift-evolution/pull/1271.